NOISSUE: add subtopic support for nats messages

beres opened the pull request:
### What does this do?
Nats listener subscribe to "channels.>" to catch all messages to manage also subtopic.

### Which issue(s) does this PR fix/relate to?

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

tttale committed:
added subtopic support from nats messages

Signed-off-by: ale <ale@beffect.net>

drasko reviewed (commented):
Looks good and very nice fix, I just need to re-check this regexp in little more detail.

@nmarcetic please review

manuio commented on mqtt/mqtt.js:
```diff
@@ -80,14 +80,16 @@ function startMqtt() {
             cmd: 'publish',
             qos: 2,
-            topic: 'channels/' + m.Channel + '/messages',
+            topic: 'channels/' + channel[1] + '/messages' + (
```
Just curious, what is channel[0] ?

beres replied:
hi,

0 is always all the string if match the regexp
1 channel id
2 the remaing part  of the string or undefined
3 the remaing of the string or undefined

as example is m.Channel is "a22de3db-61df-461a-806f-5c2ed7f4cb10.data" channel var will be
[ 'a22de3db-61df-461a-806f-5c2ed7f4cb10.data', 'a22de3db-61df-461a-806f-5c2ed7f4cb10',  '.data',  'data',  index: 0,  input: 'a22de3db-61df-461a-806f-5c2ed7f4cb10.data' ]

while if m.Channel is "a22de3db-61df-461a-806f-5c2ed7f4cb10" channel var will be
[ 'a22de3db-61df-461a-806f-5c2ed7f4cb10', 'a22de3db-61df-461a-806f-5c2ed7f4cb10',  undefined, undefined,  index: 0,  input: 'a22de3db-61df-461a-806f-5c2ed7f4cb10' ]

beres replied:
the Reg exp could be simplier if you prefer to strip manually the first dot :smile:

manuio replied:
Got it, thank you1

manuio reviewed (dismissed):
@beres LGTM. Just update the branch please

nmarcetic reviewed (commented):
@beres Maybe is a better idea if you can align it with [authorizePublish](https://github.com/mainflux/mainflux/blob/2f3fe5e5e70dbec2fd50e079f55258786ab842f4/mqtt/mqtt.js#L101) maybe even move this parse logic to separate helper function and just use it in both publish and subscribe, its exactly the same. I don't like the idea to repeat the code and do the same thing in 2 different ways, maybe I am wrong but looks like the same logic no? What do you think?

beres committed:
unify topic parsing between publish and subscribe

drasko commented on mqtt/mqtt.js:
```diff
@@ -101,29 +107,21 @@ nats.subscribe('channel.>', {'queue':'mqtts'}, function (msg) {
     // Topics are in the form `channels/<channel_id>/messages`
     // Subtopic's are in the form `channels/<channel_id>/messages/<subtopic>`
-    var channel = /^channels\/(.+?)\/messages\/?.*$/.exec(packet.topic);
+    var channel = parseMqttTopic(packet.topic)
```
This name is confusing. Regexp does not return only channel ID in the selection, but subtopic as well from what I see.

Something like `topicParts` would make more sense.

beres replied:
RegExp return and array with channel name and the topic parts, maybe i can raneme it channelParts, what do you thing?

drasko replied:
Yes, this is good also. Just not `channel`, as it is missleadning.

drasko commented on mqtt/mqtt.js:
```diff
@@ -101,29 +107,21 @@ nats.subscribe('channel.>', {'queue':'mqtts'}, function (msg) {
+
     var channelId = channel[1],
+        channelTopic = 'channel.' + channelId + (
+            channel[2] !== undefined ? channel[2].replace(/\//g, '.') : ''
```
Also add a comment here what are these hard-coded array members (what they represent).

beres committed:
add comment on publish

beres committed:
rename chennel variable in channelParts

drasko reviewed (approved):
LGTM

drasko commented on mqtt/mqtt.js:
```diff
@@ -80,14 +80,22 @@ function startMqtt() {
-        packet;
+        packet, channel;
     if (m && m.Protocol !== 'mqtt') {
+        channel = /^([\w\-]+)(\.(.+))?$/.exec(m.Channel)
```
Again, is this `channel` or `channelParts`? Looks like there are two selections here.

beres committed:
- nat.subscribe: renamed variable channel in channelParts
- nat.subsscribe: rewrite simplier regexp
- nat.subscribe: fix full topic name composition

drasko reviewed (approved):
LGTM

anovakovic01 commented on mqtt/mqtt.js:
```diff
@@ -80,14 +80,27 @@ function startMqtt() {
     if (m && m.protocol !== 'mqtt') {
+        // rexexp has two element, the first one is the channelId, wile
+        // the second is undefined or the subtopic parts starting from the dot
+        channelParts = /^([\w\-]+)(\..+[^\.]$)*$/.exec(m.channel)
```
@drasko AFAIK earlier we used `channel` only for channel ID. This means that we'll store whole topic in `channel` field, which is not only channel ID. Just want you to be aware of this. Also, it won't work with the current system (I think).

drasko replied:
Why wouldn't it?

anovakovic01 replied:
Sorry, didn't realized that there was a dot there. But still, I didn't know that you've already decided that you're going to pass subtopic here too. Maybe the rest of the team should be in sync with this change too.

drasko replied:
Yep, subtopics will be enabled for all 4 adapters. @nmarcetic - remainder to open the issues.

anovakovic01 replied:
That's OK, but there is no spec how will this be done, that's why I don't know if this is how we are going to do it. So we are going to pass `<channel_id>.subtopic` in `channel` field? Why not adding separate field that will be called subtopic, and leave `channel` field as is? Then there will be no need to extract `channel_id` from string.

drasko replied:
Yes, that would make code probably cleaner. But to see exactly what you propose, can you write here a small snippet example please.

anovakovic01 replied:
So, instead of using current `channel` field from here, I would add a new field: `subtopic`. So our structure would look like this:
```
message RawMessage {
	string channel     = 1;
        string subtopic = 2;
	string publisher   = 3;
	string protocol    = 4;
	string contentType = 5;
	bytes  payload     = 6;
}
```
In `channel`, we would store just channel ID, like we've done so far, and in `subtopic` we would store everything else (subtopic).

anovakovic01 commented on mqtt/mqtt.js:
```diff
@@ -80,14 +80,27 @@ function startMqtt() {
-            topic: 'channels/' + m.channel + '/messages',
+            // compose the topic with channelId followed by /messages and
+            // all the subtopics in the form /subtopic/subN/
+            topic: 'channels/' + channelParts[1] + '/messages' + (
```
What if `channelParts` has length 1?

drasko replied:
Good catch there!

beres replied:
@anovakovic01 afsik if the regexp match it will be always of length 3, first element contains the full string, second the channeid, and the third could be undefined or the subtopic part

anovakovic01 replied:
@beres Sorry, I missed this check later.

anovakovic01 commented on mqtt/mqtt.js:
```diff
@@ -99,29 +112,23 @@ nats.subscribe('channel.*', {'queue':'mqtts'}, function (msg) {
     }
-    var channelId = channel[1],
+
+    var channelId = channelParts[1],
```
What if `channelParts` has length 1.

anovakovic01 commented on mqtt/mqtt.js:
```diff
@@ -149,13 +156,13 @@ aedes.authorizePublish = function (client, packet, publish) {
       return;
     }
-    var channelId = channel[1],
+    var channelId = channelParts[1],
```
Same here, what if `channelParts` has length 1.
